File: fortrangl.c
#include <GL/glew.h>
#include <GLFW/glfw3.h>
#include <stdlib.h>
#include <stdio.h>
#include <ctype.h>
#include <string.h>
#include <assert.h>

#include <ISO_Fortran_binding.h>

#ifndef __DEBUG__
#define __DEBUG__ 1
#endif


extern void glfresize_func(GLFWwindow *window, int width, int height);
extern void glfkey_func(GLFWwindow *window, int key, int scancode, int action, int mods);
extern void glfcursor_func(GLFWwindow *window, double xpos, double ypos);
extern void glfmousebutton_func(GLFWwindow *window, int button, int action, int mods);

static void error_callback(int error, const char* description)
{
    fputs(description, stderr);
}

static char *f_str_to_c_str(CFI_cdesc_t *data)
{
  assert(data);
  assert(data->base_addr);
  assert(data->rank == 0);
  assert(data->type == 12);
  GLuint index = 0;
  char *fstr = data->base_addr;
  int max_len = data->elem_len;

  char *cstr = (char *)malloc(max_len + 1);

  bzero(cstr, max_len);
  for(int i=0; i<max_len && fstr[i] != 0; i++) {
    cstr[i] = fstr[i];
  }

  return cstr;
}
		    
GLFWwindow *glfInit(int width, int height)
{
  GLFWwindow *window;

  glfwSetErrorCallback(error_callback);
  if (!glfwInit()) {
    exit(EXIT_FAILURE);
  }
  
  glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);
  glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 3);
  #ifdef __APPLE__
  glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GL_TRUE);
  #endif
  glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);

  window = glfwCreateWindow(640, 480, "FortranGL", NULL, NULL);
  if (!window) {
    glfwTerminate();
    exit(EXIT_FAILURE);
  }

  glfwMakeContextCurrent(window);

  if (glewInit()) {
    exit(EXIT_FAILURE);
  }

  glfwSetWindowSizeCallback(window, glfresize_func);
  glfwSetKeyCallback(window, glfkey_func);
  glfwSetCursorPosCallback(window, glfcursor_func);
  glfwSetMouseButtonCallback(window, glfmousebutton_func);
  
  return window;
}

#if __DEBUG__
GLuint _glfError(const char *function, int line)
{
  GLenum err = glGetError();

  if (err) {
    printf("%s:%d ", function, line);
  }

  switch(err) {
  case GL_NO_ERROR:
    break;

  case GL_INVALID_ENUM:
    printf("GL_INVALID_ENUM\n");
    break;
    
  case GL_INVALID_VALUE:
    printf("GL_INVALID_VALUE\n");
    break;
    
  case GL_INVALID_OPERATION:
    printf("GL_INVALID_OPERATION\n");
    break;
    
  case GL_INVALID_FRAMEBUFFER_OPERATION:
    printf("GL_INVALID_FRAMEBUFFER_OPERATION\n");
    break;
    
  case GL_OUT_OF_MEMORY:
    printf("GL_OUT_OF_MEMORY\n");
    break;

  default:
    assert(0);
  }

  return err;
}
#else
#define _glfError(function, line)
#endif

void glfClearBuffers(GLint color, GLint depth, GLint stencil)
{
  GLuint mask = 0;
  
  if (color)
    mask |= GL_COLOR_BUFFER_BIT;
  
  if (depth)
    mask |= GL_DEPTH_BUFFER_BIT;
  
  if (stencil)
    mask |= GL_STENCIL_BUFFER_BIT;

  glClear(mask);
}



/****************************************************************************************/
/* Array Buffer support */
/****************************************************************************************/

GLuint glfGenArrayBuffer()
{
  GLuint buffer = 0;

  glGenBuffers(1, &buffer);

  return buffer;
}

void glfBindArrayBuffer(GLuint buffer)
{
  glBindBuffer(GL_ARRAY_BUFFER, buffer);
}

void glfArrayBufferData(CFI_cdesc_t *data)
{
  assert(data);
  assert(data->dim);
  size_t len = data->elem_len;
  len *= data->dim->extent;
  
  glBufferData(GL_ARRAY_BUFFER, len, data->base_addr, GL_STATIC_DRAW);
}

void glfArrayBufferSubData(size_t offset, CFI_cdesc_t *data)
{
  assert(data);
  assert(data->dim);
  size_t len = data->elem_len;
  len *= data->dim->extent;
    
  offset *= data->elem_len;

  glBufferSubData(GL_ARRAY_BUFFER, len, offset, data->base_addr);
}



/****************************************************************************************/
/* Element Buffer support */
/****************************************************************************************/

GLuint glfGenElementBuffer()
{
  GLuint buffer = 0;

  glGenBuffers(1, &buffer);

  return buffer;
}

void glfBindElementBuffer(GLuint buffer)
{
  glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, buffer);
}

void glfElementBufferData(CFI_cdesc_t *data)
{
  assert(data);
  assert(data->dim);
  size_t len = data->elem_len;
  len *= data->dim->extent;
        
  glBufferData(GL_ELEMENT_ARRAY_BUFFER, len, data->base_addr, GL_STATIC_DRAW);
}

void glfElementBufferSubData(size_t offset, CFI_cdesc_t *data)
{
  assert(data);
  assert(data->dim);
  size_t len = data->elem_len;
  len *= data->dim->extent;
    
  offset *= data->dim->extent;
  
  glBufferSubData(GL_ELEMENT_ARRAY_BUFFER, len, offset, data->base_addr);
}


/****************************************************************************************/
/* Uniform Buffer support */
/****************************************************************************************/
GLuint glfGenUniformBuffer()
{
  GLuint buffer = 0;

  glGenBuffers(1, &buffer);

  return buffer;
}

void glfBindUniformBuffer(GLuint buffer)
{
  glBindBuffer(GL_UNIFORM_BUFFER, buffer);
}

void glfUniformBufferData(CFI_cdesc_t *data)
{
  assert(data);
  assert(data->dim);
  size_t len = data->elem_len;
  len *= data->dim->extent;
  
  glBufferData(GL_UNIFORM_BUFFER, len, data->base_addr, GL_STATIC_DRAW);
}

void glfUniformBufferSubData(size_t offset, CFI_cdesc_t *data)
{
  assert(data);
  assert(data->dim);
  size_t len = data->elem_len;
  len *= data->dim->extent;
    
  offset *= data->elem_len;

  glBufferSubData(GL_UNIFORM_BUFFER, len, offset, data->base_addr);
}


void glfBindUniformBufferBase(GLuint index, GLuint buffer)
{
  glBindBufferBase(GL_UNIFORM_BUFFER, index, buffer);
}


GLuint glfUniformBlockIndex(GLuint program, CFI_cdesc_t *data)
{
  char *cstr = f_str_to_c_str(data);
  GLuint index;
  
  index = glGetUniformBlockIndex(program, cstr);

  free(cstr);
  
  return index;
}

GLuint glfUniformMaxBufferBindings()
{
  GLuint count = 0;

  glGetIntegerv(GL_MAX_UNIFORM_BUFFER_BINDINGS, &count);

  return count;
}


/****************************************************************************************/
/* Vertex Array Buffer support */
/****************************************************************************************/
GLuint glfGenVAO()
{
  GLuint buffer = 0;

  glGenVertexArrays(1, &buffer);

  return buffer;
}

void glfBindVAO(GLuint buffer)
{
  glBindVertexArray(buffer);
}


/****************************************************************************************/
/* Shader compilation routines */
/****************************************************************************************/
static GLuint _glfCompileShader(const char *vertex_str, const char *fragment_str)
{
  // Create an empty vertex shader handle
  GLuint vertexShader = glCreateShader(GL_VERTEX_SHADER);

  // Send the vertex shader source code to GL
  glShaderSource(vertexShader, 1, &vertex_str, 0);

  // Compile the vertex shader
  glCompileShader(vertexShader);

  GLint isCompiled = 0;
  glGetShaderiv(vertexShader, GL_COMPILE_STATUS, &isCompiled);
  if(isCompiled == GL_FALSE) {
    GLint maxLength = 0;
    glGetShaderiv(vertexShader, GL_INFO_LOG_LENGTH, &maxLength);

    // The maxLength includes the NULL character
    char *error = (char *)malloc(maxLength);
    glGetShaderInfoLog(vertexShader, maxLength, &maxLength, error);

    // We don't need the shader anymore.
    glDeleteShader(vertexShader);

    // Use the infoLog as you see fit.
    printf("Error compiling the vertex shader: %s\n", error);

    // In this simple program, we'll just leave
    return 0;
  }

  // Create an empty fragment shader handle
  GLuint fragmentShader = glCreateShader(GL_FRAGMENT_SHADER);

  // Send the fragment shader source code to GL
  glShaderSource(fragmentShader, 1, &fragment_str, 0);

  // Compile the fragment shader
  glCompileShader(fragmentShader);

  glGetShaderiv(fragmentShader, GL_COMPILE_STATUS, &isCompiled);
  if(isCompiled == GL_FALSE) {
    GLint maxLength = 0;
    glGetShaderiv(fragmentShader, GL_INFO_LOG_LENGTH, &maxLength);

    // The maxLength includes the NULL character
    char *error = (char *)malloc(maxLength);
    glGetShaderInfoLog(fragmentShader, maxLength, &maxLength, error);

    // We don't need the shader anymore.
    glDeleteShader(fragmentShader);
    glDeleteShader(vertexShader);

    // Use the infoLog as you see fit.
    printf("Error compiling the fragment shader: %s\n", error);

    // In this simple program, we'll just leave
    return 0;
  }

  // Vertex and fragment shaders are successfully compiled.
  // Now time to link them together into a program.
  // Get a program object.
  GLuint program = glCreateProgram();

  // Attach our shaders to our program
  glAttachShader(program, vertexShader);
  glAttachShader(program, fragmentShader);

  // Link our program
  glLinkProgram(program);

  // Note the different functions here: glGetProgram* instead of glGetShader*.
  GLint isLinked = 0;
  glGetProgramiv(program, GL_LINK_STATUS, (int *)&isLinked);
  if (isLinked == GL_FALSE) {
    GLint maxLength = 0;
    glGetProgramiv(program, GL_INFO_LOG_LENGTH, &maxLength);

    // The maxLength includes the NULL character
    char *error = (char *)malloc(maxLength);
    glGetProgramInfoLog(program, maxLength, &maxLength, error);

    // We don't need the program anymore.
    glDeleteProgram(program);
    // Don't leak shaders either.
    glDeleteShader(vertexShader);
    glDeleteShader(fragmentShader);

    // Use the infoLog as you see fit.

    // In this simple program, we'll just leave
    return 0;
  }

  // Always detach shaders after a successful link.
  glDetachShader(program, vertexShader);
  glDetachShader(program, fragmentShader);

  return program;
}

/****************************************************************************************/
/* GLSL program support */
/****************************************************************************************/

GLuint glfCompileShaders(CFI_cdesc_t *vdata, CFI_cdesc_t *fdata)
{
  char *vertex_shader = f_str_to_c_str(vdata);
  char *fragment_shader = f_str_to_c_str(fdata);
  GLuint program;

  program = _glfCompileShader(vertex_shader, fragment_shader);

  if (program == 0) {
    printf("GLSL compilation failed\n\n%s\n\n%s\n", vertex_shader, fragment_shader);
  }

  printf("%s\n",vertex_shader);
  printf("%s\n",fragment_shader);
  
  free(vertex_shader);
  free(fragment_shader);
  
  if (program) {
      glUseProgram(program);
  }
  
  return program;  
}


/****************************************************************************************/
/* Vertex binding support PC */
/****************************************************************************************/
GLuint glfBindVertexAttribPointers(CFI_cdesc_t *data)
{
  char *vertex_desc_str = f_str_to_c_str(data);
  char *str = vertex_desc_str;
  GLuint types[32], counts[32];
  size_t offsets[32];
  GLuint max_attrib = 0;
  GLuint vertex_size = 0;
  
  if (strlen(vertex_desc_str) == 0)
    return 1;
      
  for(int i=0; *str; i++) {
    GLuint type_size = 0;
    GLuint type_count = 0;

    offsets[i] = vertex_size;
    
    switch(toupper(*str)) {
    case 'F':
      types[i] = GL_FLOAT;
      type_size = 4;
      break;
      
    case 'D':
      types[i] = GL_DOUBLE;
      type_size = 8;
      break;
      
    case 'I':
      types[i] = GL_INT;
      type_size = 4;
      break;

    default:
      goto error_return;
    }

    if (*str == 0)
      goto error_return;
    
    str++;

    type_count = *str++ - '0';
    counts[i] = type_count;

    vertex_size += type_size * type_count;
    
    max_attrib++;
  }
    
  printf("Vertex Desc %s vertex_size %d attrib_count %d\n", vertex_desc_str, vertex_size, max_attrib);

  for(int i=0; i<max_attrib; i++) {
    if (types[i] == GL_FLOAT)
      glVertexAttribPointer(i, counts[i], GL_FLOAT, GL_FALSE, vertex_size, (void *)offsets[i]);
    else if (types[i] == GL_INT)
      glVertexAttribIPointer(i, counts[i], GL_INT, vertex_size, (void *)offsets[i]);
    else if (types[i] == GL_DOUBLE)
      glVertexAttribLPointer(i, counts[i], GL_DOUBLE, vertex_size, (void *)offsets[i]);      
    else
      assert(0);
    
    glEnableVertexAttribArray(i);
  }

  // disable unused attribs
  GLuint max_gl_attribs = 0;
  glGetIntegerv(GL_MAX_VERTEX_ATTRIBS, &max_gl_attribs);

  for(int i=max_attrib; i<max_gl_attribs; i++) {
    glDisableVertexAttribArray(i);
  }
    
  free(vertex_desc_str);
  
  return 0;

 error_return:
  free(vertex_desc_str);
  
  return 1;
}


/****************************************************************************************/
/* Draw calls */
/****************************************************************************************/
void glfDrawTriangleArrays(GLuint start, GLuint count)
{
  glDrawArrays(GL_TRIANGLES, start, count);
}

File: fortrangl.f90
!
!  Created by Michael Larson
!  January 13 2020
!
!  I don't care how this is used and I am not repsonsible for it's use, just keep
!  my name at the top of the source.
!
!  Thanks
!
!  Michael Larson
!

!****************************************************************************************
! GLSL Types bucket
!****************************************************************************************
module glsl_types
  use, intrinsic :: iso_c_binding
  implicit none

  real(c_float),parameter :: PI = 3.14159

  type, bind(C) :: Vec4
     real(c_float),dimension(1:4) :: v
  end type Vec4

  type, bind(C) :: Mat4
     real(c_float),dimension(1:4,1:4) :: m
  end type Mat4

contains
  function deg2rad(angle)
    use, intrinsic :: iso_c_binding
    real(c_float),intent(in),value :: angle
    real(c_float) :: deg2rad

    deg2rad = angle * PI / 180.0
  end function deg2rad
end module glsl_types



module glf
  use, intrinsic :: iso_c_binding
  use glsl_types
  implicit none

  ! A Vertex is a derived type that the programmer is required to fill in if
  ! they want to use something other than Position / Color information
  type, bind(C) :: Vertex
     real(c_float) :: x, y, z, w
     real(c_float) :: r, g, b, a
  end type Vertex

  integer,parameter :: GL_POINTS = 0
  integer,parameter :: GL_LINES = 1
  integer,parameter :: GL_LINE_LOOP = 2
  integer,parameter :: GL_LINE_STRIP = 3
  integer,parameter :: GL_TRIANGLES = 4
  integer,parameter :: GL_TRIANGLE_STRIP = 5
  integer,parameter :: GL_TRIANGLE_FAN = 6

  integer,parameter :: GL_DEPTH_TEST = 2929
  
  interface
     ! *********************************************************************************
     ! glf window calls
     ! *********************************************************************************
     function glfInit(width, height, major, minor, window_name) bind(C, name="glfInit")
       use, intrinsic :: iso_c_binding
       type(c_ptr) :: glfInit
       integer(c_int),intent(in),value :: width, height, major, minor
       character(len=*),intent(in) :: window_name
     end function glfInit

     subroutine glfSetCurrentContext(window) bind(C, name="glfwMakeContextCurrent")
       use, intrinsic :: iso_c_binding
       type(c_ptr),intent(in),value :: window
     end subroutine glfSetCurrentContext

     integer function glfWindowShouldClose(window) bind(C, name="glfwWindowShouldClose")
       use, intrinsic :: iso_c_binding
       type(c_ptr),intent(in),value :: window
     end function glfWindowShouldClose

     subroutine glfGetWindowSize(window, pWidth, pHeight) bind(C, name="glfwGetWindowSize")
       use, intrinsic :: iso_c_binding
       type(c_ptr),intent(in),value :: window
       integer,pointer :: pWidth, pHeight
     end subroutine glfGetWindowSize

     subroutine glfDestroyWindow(window) bind(C, name="glfwDestroyWindow")
       use, intrinsic :: iso_c_binding
       type(c_ptr),intent(in),value :: window
     end subroutine glfDestroyWindow

     subroutine glfSwapBuffers(window) bind(C, name="glfwSwapBuffers")
       use, intrinsic :: iso_c_binding
       type(c_ptr),intent(in),value :: window
     end subroutine glfSwapBuffers

     subroutine glfPollEvents() bind(C, name="glfwPollEvents")
     end subroutine glfPollEvents



     ! *********************************************************************************
     ! native opengl calls
     ! *********************************************************************************
     subroutine glfViewport(x, y, width, height) bind(C, name="glViewport")
       use, intrinsic :: iso_c_binding
       integer,intent(in),value :: x, y, width, height
     end subroutine glfViewport

     subroutine glfClearColor(red, green, blue, alpha) bind(C, name="glClearColor")
       use, intrinsic :: iso_c_binding
       real(c_float),intent(in),value :: red, green, blue, alpha
     end subroutine glfClearColor

     subroutine glfClearBuffers(color, depth, stencil) bind(C, name="glfClearBuffers")
       use, intrinsic :: iso_c_binding
       integer,intent(in),value :: color, depth, stencil
     end subroutine glfClearBuffers

     subroutine glfEnable(enum) bind(C, name="glfEnable")
       use, intrinsic :: iso_c_binding
       integer(c_int),intent(in),value :: enum
     end subroutine glfEnable
     
     subroutine glfDisable(enum) bind(C, name="glfDisable")
       use, intrinsic :: iso_c_binding
       integer(c_int),intent(in),value :: enum
     end subroutine glfDisable
     
       


     ! *********************************************************************************
     ! misc glf calls
     ! *********************************************************************************
     subroutine glfError(line) bind(C, name="glfError")
       use, intrinsic :: iso_c_binding
       integer,intent(in),value :: line
     end subroutine glfError


     ! *********************************************************************************
     ! glf array buffer calls
     ! *********************************************************************************
     function glfGenArrayBuffer() bind(C, name="glfGenArrayBuffer")
       use, intrinsic :: iso_c_binding
       integer(c_int) :: glfGenArrayBuffer
     end function glfGenArrayBuffer

     subroutine glfBindArrayBuffer(buffer) bind(C, name="glfBindArrayBuffer")
       use, intrinsic :: iso_c_binding
       integer(c_int),intent(in),value :: buffer
     end subroutine glfBindArrayBuffer

     subroutine glfArrayBufferData(data) bind(C, name="glfArrayBufferData")
       use, intrinsic :: iso_c_binding
       import Vertex
       type(Vertex),pointer,dimension(:),intent(in) :: data
     end subroutine glfArrayBufferData

     subroutine glfArrayBufferSubData(offset,data) bind(C, name="glfArrayBufferSubData")
       use, intrinsic :: iso_c_binding
       integer(c_long),intent(in),value :: offset
       type(c_ptr),intent(in) :: data
     end subroutine glfArrayBufferSubData



     ! *********************************************************************************
     ! glf element buffer calls
     ! *********************************************************************************
     function glfGenElementBuffer() bind(C, name="glfGenElementBuffer")
       use, intrinsic :: iso_c_binding
       integer(c_int) :: glfGenElementBuffer
     end function glfGenElementBuffer

     subroutine glfBindElementBuffer(buffer) bind(C, name="glfBindElementBuffer")
       use, intrinsic :: iso_c_binding
       integer(c_int),intent(in),value :: buffer
     end subroutine glfBindElementBuffer

     subroutine glfElementBufferData(data) bind(C, name="glfElementBufferData")
       use, intrinsic :: iso_c_binding
       type(c_ptr),intent(in) :: data
     end subroutine glfElementBufferData

     subroutine glfElementBufferSubData(offset,data) bind(C, name="glfElementBufferSubData")
       use, intrinsic :: iso_c_binding
       integer(c_long),intent(in),value :: offset
       type(c_ptr),intent(in) :: data
     end subroutine glfElementBufferSubData



     ! *********************************************************************************
     ! glf uniform buffer calls
     ! *********************************************************************************
     function glfGenUniformBuffer() bind(C, name="glfGenUniformBuffer")
       use, intrinsic :: iso_c_binding
       integer(c_int) :: glfGenUniformBuffer
     end function glfGenUniformBuffer

     subroutine glfBindUniformBuffer(buffer) bind(C, name="glfBindUniformBuffer")
       use, intrinsic :: iso_c_binding
       integer(c_int),intent(in),value :: buffer
     end subroutine glfBindUniformBuffer

     subroutine glfUniformBufferData(data) bind(C, name="glfUniformBufferData")
       use, intrinsic :: iso_c_binding
       real(c_float),pointer,dimension(:),intent(in) :: data
     end subroutine glfUniformBufferData

     subroutine glfUniformBufferSubData(offset,data) bind(C, name="glfUniformBufferSubData")
       use, intrinsic :: iso_c_binding
       integer(c_long),intent(in),value :: offset
       real(c_float),intent(in) :: data
     end subroutine glfUniformBufferSubData

     subroutine glfUniformBindBufferBase(index, buffer) bind(C, name="glfUniformBindBufferBase")
       use, intrinsic :: iso_c_binding
       integer(c_int),intent(in),value :: index, buffer
     end subroutine glfUniformBindBufferBase

     function glfUniformBlockIndex(glsl_program, uniform_block_str) bind(C, name="glfUniformBlockIndex")
       use, intrinsic :: iso_c_binding
       integer(c_int),intent(in),value :: glsl_program 
       character(len=*),intent(in) :: uniform_block_str
       integer(c_int) :: glfUniformBlockIndex
     end function glfUniformBlockIndex

     subroutine glfUniformBlockBinding(glsl_program, index, binding) bind(C, name="glUniformBlockBinding")
       use, intrinsic :: iso_c_binding
       integer(c_int),intent(in),value :: glsl_program, index, binding
     end subroutine glfUniformBlockBinding

     function glfUniformMaxBufferBindings() bind(C, name="glfUniformMaxBufferBindings")
       use, intrinsic :: iso_c_binding
       integer(c_int) :: glfUniformMaxBufferBindings
     end function glfUniformMaxBufferBindings



     ! *********************************************************************************
     ! glf vertex array buffer calls
     ! *********************************************************************************
     function glfGenVAO() bind(C, name="glfGenVAO")
       use, intrinsic :: iso_c_binding
       integer(c_int) :: glfGenVAO
     end function glfGenVAO

     subroutine glfBindVAO(buffer) bind(C, name="glfBindVAO")
       use, intrinsic :: iso_c_binding
       integer(c_int),intent(in),value :: buffer
     end subroutine glfBindVAO



     ! *********************************************************************************
     ! GLSL program calls
     ! *********************************************************************************
     function glfCompileShaderFiles(vertex_filename, fragment_filename) bind(C, name="glfCompileShaderFiles")
       use, intrinsic :: iso_c_binding
       character(len=*),intent(in) :: vertex_filename, fragment_filename
       integer(c_int) :: glfCompileShaderFiles
     end function glfCompileShaderFiles

     ! don't use.... shader strings from fortran to C are a real pain to deal with...
     ! eventually I will write a simple compiler to save these into a __DATA__ section as
     ! strings from file which will make this part of the compilation
     function glfCompileShaders(vertex_str, fragment_str) bind(C, name="glfCompileShaders")
       use, intrinsic :: iso_c_binding
       character(len=*),intent(in) :: vertex_str, fragment_str
       integer(c_int) :: glfCompileShaders
     end function glfCompileShaders



     ! *********************************************************************************
     ! Vertex Descriptor support for vertex attribs
     ! *********************************************************************************
     function glfBindVertexAttribPointers(vertex_desc_str) bind(C, name="glfBindVertexAttribPointers")
       use, intrinsic :: iso_c_binding
       character(len=*),intent(in) :: vertex_desc_str
       integer(c_int) :: glfBindVertexAttribPointers
     end function glfBindVertexAttribPointers



     ! *********************************************************************************
     ! Position Color Vertex support, default format
     ! *********************************************************************************
     function glfBindDefaultPCShaders() bind(C, name="glfBindDefaultPCShaders")
       use, intrinsic :: iso_c_binding
       integer(c_int) :: glfBindDefaultPCShaders
     end function glfBindDefaultPCShaders

     function glfBindPCVertexAttribPointers() bind(C, name="glfBindPCVertexAttribPointers")
       use, intrinsic :: iso_c_binding
       integer(c_int) :: glfBindPCVertexAttribPointers
     end function glfBindPCVertexAttribPointers



     ! *********************************************************************************
     ! glf draw calls
     ! *********************************************************************************
     subroutine glfDrawArrays(mode, offset, count) bind(C, name="glfDrawArrays")
       use, intrinsic :: iso_c_binding
       integer(c_int),intent(in),value :: mode, offset, count
     end subroutine glfDrawArrays

     subroutine glfDrawElements(mode, offset, count, indices) bind(C, name="glfDrawElements")
       use, intrinsic :: iso_c_binding
       integer(c_int),intent(in),value :: mode, offset, count
       integer(c_int),pointer,dimension(:),intent(in) :: indices
     end subroutine glfDrawElements


     
     ! *********************************************************************************
     ! glf GLU calls
     ! *********************************************************************************
     subroutine glfPerspective(mat, fovy, aspect, near, far) bind(C, name="glfPerspective")
       use, intrinsic :: iso_c_binding
       import Mat4
       type(Mat4),intent(inout),pointer :: mat
       real(c_float),intent(in),value :: fovy, aspect, near, far
     end subroutine glfPerspective

     subroutine glfLookAt(mat, eye, center, up) bind(C, name="glfLookAt")
       use, intrinsic :: iso_c_binding
       import Mat4, Vec4
       type(Mat4),intent(inout),pointer :: mat
       type(Vec4),intent(in),pointer :: eye, center, up
     end subroutine glfLookAt

     subroutine glfCalcMVP(mvp, model, view, projection) bind(C, name="glfCalcMVP")
       use, intrinsic :: iso_c_binding
       import Mat4, Vec4
       type(Mat4),intent(inout),pointer :: mvp
       type(Mat4),intent(in),pointer :: model, view, projection
     end subroutine glfCalcMVP

     
     subroutine glfIdentity(mat) bind(C, name="glfIdentity")
       use, intrinsic :: iso_c_binding
       import Mat4
       type(Mat4),intent(inout),pointer :: mat
     end subroutine glfIdentity
       
     subroutine glfRotate(mat, angle, x, y, z) bind(C, name="glfRotate")
       use, intrinsic :: iso_c_binding
       import Mat4
       type(Mat4),intent(inout),pointer :: mat
       real(c_float),intent(in),value :: angle
       real(c_float),intent(in),value :: x, y, z
     end subroutine glfRotate
       
     subroutine glfTranslate(mat, x, y, z) bind(C, name="glfTranslate")
       use, intrinsic :: iso_c_binding
       import Mat4
       type(Mat4),intent(inout),pointer :: mat
       real(c_float),intent(in),value :: x, y, z
     end subroutine glfTranslate
       
     subroutine glfScale(mat, x, y, z) bind(C, name="glfScale")
       use, intrinsic :: iso_c_binding
       import Mat4
       type(Mat4),intent(inout),pointer :: mat
       real(c_float),intent(in),value :: x, y, z
     end subroutine glfScale

     subroutine glfDumpFramebufferToTGA(window, filename) bind(C, name="glfDumpFramebufferToTGA")
       use, intrinsic :: iso_c_binding
       import Mat4
       type(c_ptr),intent(in),value :: window
       character(len=*),intent(in) :: filename
     end subroutine glfDumpFramebufferToTGA
     
    end interface


contains
  subroutine glfSetVertexPos(buffer, index, x, y, z, w)
    use, intrinsic :: iso_c_binding
    type(Vertex),pointer,dimension(:) :: buffer
    integer :: index
    real(c_float) :: x, y, z, w

    buffer(index)%x = x
    buffer(index)%y = y
    buffer(index)%z = z
    buffer(index)%w = w
  end subroutine glfSetVertexPos

  subroutine glfSetVertexCol(buffer, index, r, g, b, a)
    use, intrinsic :: iso_c_binding
    type(Vertex),pointer,dimension(:) :: buffer
    integer :: index
    real(c_float) :: r, g, b, a

    buffer(index)%r = r
    buffer(index)%g = g
    buffer(index)%b = b
    buffer(index)%a = a
  end subroutine glfSetVertexCol

  ! *********************************************************************************
  ! glf event handler calls
  ! *********************************************************************************
  subroutine glfresize_func(window, width, height) bind(C, name="glfresize_func")
    use,intrinsic :: iso_c_binding
    type(c_ptr),intent(in),pointer :: window
    integer(c_int),intent(in),value :: width, height

    print *, width, height

    call glfViewport(0, 0, width, height)
  end subroutine glfresize_func

  subroutine glfkey_func(window, key, scancode, action, mods) bind(C, name="glfkey_func")
    use,intrinsic :: iso_c_binding
    type(c_ptr),intent(in),pointer :: window
    integer(c_double),intent(in),value :: key, scancode, action, mods

    !if (key == GLFW_KEY_ESCAPE && action == GLFW_PRESS) then
    !  glfSetWindowShouldClose(window);
    !end if
  end subroutine glfkey_func

  subroutine glfcursor_func(window, xpos, ypos) bind(C, name="glfcursor_func")
    use,intrinsic :: iso_c_binding
    type(c_ptr),intent(in),pointer :: window
    integer(c_double),intent(in),value :: xpos, ypos

    !print *, xpos, ypos
  end subroutine glfcursor_func

  subroutine glfmousebutton_func(window, button, action, mods) bind(C, name="glfmousebutton_func")
    use,intrinsic :: iso_c_binding
    type(c_ptr),intent(in),pointer :: window
    integer(c_int),intent(in),value :: button, action, mods

    !print *, button, action, mods
  end subroutine glfmousebutton_func
end module glf


